tests/test_runner/tests.py

```diff
@@ -181,6 +181,19 @@ def test_reorder_test_bin_random_and_reverse(self):
             'Tests1.test1', 'Tests1.test2', 'Tests2.test2', 'Tests2.test1',
         ])
 
+    def test_reorder_tests_same_type_consecutive(self):
+        """Tests of the same type are made consecutive."""
+        tests = self.make_tests()
+        # Move the last item to the front.
+        tests.insert(0, tests.pop())
+        self.assertTestNames(tests, expected=[
+            'Tests2.test2', 'Tests1.test1', 'Tests1.test2', 'Tests2.test1',
+        ])
+        reordered_tests = reorder_tests(tests, classes=[])
+        self.assertTestNames(reordered_tests, expected=[
+            'Tests2.test2', 'Tests2.test1', 'Tests1.test1', 'Tests1.test2',
+        ])
```

Comment:
I noticed that if you do the same with `shuffle` than the last test is ignored:
```console
$ ./runtests.py lookup.tests.LookupQueryingTests.test_filter_lookup_lhs lookup.tests.LookupTests.test_exact_sliced_queryset_limit_one lookup.tests.LookupQueryingTests.test_filter_wrapped_lookup_lhs  -v2 --parallel=1 --shuffle
Testing against Django installed in '/django/django'
Importing application lookup
Using shuffle seed: 5261940290 (generated)
Found 2 test(s).
...
System check identified no issues (0 silenced).
test_filter_wrapped_lookup_lhs (lookup.tests.LookupQueryingTests) ... ok
test_exact_sliced_queryset_limit_one (lookup.tests.LookupTests) ... ok

----------------------------------------------------------------------
Ran 2 tests in 0.006s

OK
Used shuffle seed: 5261940290 (generated)
Destroying test database for alias 'default' ('file:memorydb_default?mode=memory&cache=shared')...
```
vs.
```console
./runtests.py lookup.tests.LookupQueryingTests.test_filter_lookup_lhs lookup.tests.LookupTests.test_exact_sliced_queryset_limit_one lookup.tests.LookupQueryingTests.test_filter_wrapped_lookup_lhs  -v2 --parallel=1
Testing against Django installed in '/django/django'
Importing application lookup
Found 3 test(s).
...
System check identified no issues (0 silenced).
test_filter_lookup_lhs (lookup.tests.LookupQueryingTests) ... ok
test_exact_sliced_queryset_limit_one (lookup.tests.LookupTests) ... ok
test_filter_wrapped_lookup_lhs (lookup.tests.LookupQueryingTests) ... ok

----------------------------------------------------------------------
Ran 3 tests in 0.007s

OK
Destroying test database for alias 'default' ('file:memorydb_default?mode=memory&cache=shared')...
```

I added test for this case.

Reply:
> I noticed that if you do the same with shuffle than the last test is ignored:

Yes, I had noticed that, too. Thanks for adding the additional test. I was thinking this should be fixed even before `--shuffle`, though. For example, it seems like it could could result in `setUpClass()` being run twice if tests are being run in parallel.